pkg/common/idutil/spiffeid.go

```diff
@@ -0,0 +1,189 @@
+package idutil
+
+import (
+	"errors"
+	"fmt"
+	"net/url"
+	"strings"
+)
+
+type idType int
+
+const (
+	anyType idType = iota
+	trustDomainId
+	workloadId
+)
+
+// ValidateSpiffeID validates the SPIFFE ID according to the SPIFFE
+// specification. The validation mode controls the type of validation.
+func ValidateSpiffeID(spiffeID string, mode ValidationMode) error {
+	_, err := ParseSpiffeID(spiffeID, mode)
+	return err
+}
+
+// ValidateSpiffeIDURL validates the SPIFFE ID according to the SPIFFE
+// specification, namely:
+// - spiffe id is not empty
+// - spiffe id is a valid url
+// - scheme is 'spiffe'
+// - user info is not allowed
+// - host is not empty
+// - port is not allowed
+// - query values are not allowed
+// - fragment is not allowed
+// - path does not start with '/spire' since it is reserved for agent, server, etc.
+// In addition, the validation mode is used to control what kind of SPIFFE ID
+// is expected.
+// For more information:
+// [https://github.com/spiffe/spiffe/blob/master/standards/SPIFFE-ID.md]
+func ValidateSpiffeIDURL(id *url.URL, mode ValidationMode) error {
+	options := mode.validationOptions()
+
+	var kind string
+	switch {
+	case options.idType == workloadId:
+		kind = "workload "
+	case options.idType == trustDomainId:
+		kind = "trust domain "
+	}
+
+	validationError := func(format string, args ...interface{}) error {
+		return fmt.Errorf("%q is not a valid %sSPIFFE ID: "+format,
+			append([]interface{}{id.String(), kind}, args...)...)
```

Comment:
nit (continuation): Then, here you should update this line:
```go
append([]interface{}{id.String(), strings.TrimPrefix(options.idType.kind()+" ", " ")}, args...)...)
```

Reply:
I find that construction less readable. In general I agree that providing a method on the idType that returns the string representation would be desirable. In this instance though it feels a bit premature because:
1) It is only used in one spot in the code
2) The spot that uses it is only interested in special casing two of the three types and also needs to blend the string with formatting (i.e. the trailing space).

Reply:
Your comment did make me think about moving the switch inside of the validationError function, which i think tightens the scope down a bit.

Reply:
Cool!